lib/puppet/util/pidlock.rb

```diff
@@ -58,11 +69,20 @@ def clear_if_stale
     end
 
     # Ensure the process associated with this pid is our process. If
-    # not, we can unlock the lockfile. For now this is only done on
-    # POSIX and Windows platforms (PUP-9247).
+    # not, we can unlock the lockfile. CLI arguments used for identifying
+    # on POSIX depend on the os and sometimes even version.
     if Puppet.features.posix?
-      procname = Puppet::Util::Execution.execute(["ps", "-p", lock_pid, "-o", "comm="]).strip
-      args     = Puppet::Util::Execution.execute(["ps", "-p", lock_pid, "-o", "args="]).strip
+      ps_argument = ps_argument_for_current_kernel
+
+      # Check, obtain and use the right ps argument
+      begin
+        procname = Puppet::Util::Execution.execute(["ps", ps_argument, lock_pid, "-o", "comm="]).strip
+      rescue Puppet::ExecutionFailure
+        ps_argument = "-p"
+        procname = Puppet::Util::Execution.execute(["ps", ps_argument, lock_pid, "-o", "comm="]).strip
```

Comment:
If `ps_argument` was already `-p` then we end up calling `ps -p` again, which will raise and that is not rescued. But that matches the old behavior so I guess that's ok? Curious why the fallback is necessary?

Reply:
The fallback is intended only for Linux operating systems where `ps` is outdated and it does not have the `-q` parameter (such as Red Hat 5). This seemed a simpler and more future proof approach than dealing with each os and version cases.
The `ps_argument_for_current_kernel` method returns `-p` for the other kernels (Solaris, Darwin etc.) where `-p` works as intended. The old behaviour is left in place in case something changes (or was missed) for this scenario.

Reply:
Ah yeah, that makes sense. Could you amend the git commit with this info? It'll help the next person who comes along and wants to understand why it was done this way.

Reply:
Done 😃